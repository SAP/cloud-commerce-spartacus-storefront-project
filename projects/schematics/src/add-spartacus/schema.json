{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "baseSite": {
      "type": "string",
      "description": "The base site to use with spartacus",
      "default": "electronics"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application.",
      "default": "1.1"
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite whole app.component.html file.",
      "default": false
    }
  },
  "required": []
}
