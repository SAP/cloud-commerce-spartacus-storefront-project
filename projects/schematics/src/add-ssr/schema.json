{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    }
  },
  "required": []
}
