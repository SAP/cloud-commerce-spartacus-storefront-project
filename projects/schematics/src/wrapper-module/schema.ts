import { ExecutionOptions } from '@angular-devkit/schematics';

export interface Schema extends Partial<ExecutionOptions> {
  /**
   * The name of the project in which to execute schematics.
   */
  project: string;
  /**
   * When enabled, prints the additional logs.
   */
  debug?: boolean;
  /**
   * The marker module for which
   * to find or generate a new wrapper module.
   */
  markerModuleName: string;
  /**
   * Name of the feature module
   * to add to the wrapper module.
   */
  featureModuleName: string;
  /**
   * Internal options.
   * Should not be set by the user.
   */
  internal?: {
    /**
     * The execution sequence.
     * Used for ensuring one log message is printed.
     */
    sequence?: number;
  };
}
