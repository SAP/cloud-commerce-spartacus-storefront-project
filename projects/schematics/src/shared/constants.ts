export const UTF_8 = 'utf-8';

/***** Imports start *****/
export const ANGULAR_SCHEMATICS = '@schematics/angular';
export const ANGULAR_CORE = '@angular/core';
export const ANGULAR_FORMS = '@angular/forms';
export const ANGULAR_ROUTER = '@angular/router';
export const RXJS = 'rxjs';

export const SPARTACUS_CORE = '@spartacus/core';
export const SPARTACUS_STOREFRONTLIB = '@spartacus/storefront';
export const SPARTACUS_CDS = '@spartacus/cds';

export const NGRX_STORE = '@ngrx/store';
/***** Imports end *****/

/***** Classes start *****/
export const OBSERVABLE_CLASS = 'Observable';

export const STORE = 'Store';

export const FORM_BUILDER = 'FormBuilder';

export const CHANGE_DETECTOR_REF = 'ChangeDetectorRef';

export const ACTIVATED_ROUTE = 'ActivatedRoute';
export const ROUTER = 'Router';

export const CMS_COMPONENT_DATA_CLASS = 'CmsComponentData';
export const CONFIG_MODULE_CLASS = 'ConfigModule';
export const CMS_CONFIG = 'CmsConfig';
export const CMS_SELECTORS = 'CmsSelectors';
export const CMS_ACTIONS = 'CmsActions';
export const LOAD_CMS_COMPONENT_CLASS = 'LoadCmsComponent';
export const LOAD_CMS_COMPONENT_FAIL_CLASS = 'LoadCmsComponentFail';
export const LOAD_CMS_COMPONENT_SUCCESS_CLASS = 'LoadCmsComponentSuccess';
export const CMS_GET_COMPONENT_FROM_PAGE = 'CmsGetComponentFromPage';
export const USER_ADDRESS_SERVICE = 'UserAddressService';
export const AUTH_SERVICE = 'AuthService';
export const FEATURE_CONFIG_SERVICE = 'FeatureConfigService';
export const PAGE_META_RESOLVER = 'PageMetaResolver';
export const CMS_SERVICE = 'CmsService';
export const PAGE_META_SERVICE = 'PageMetaService';
export const CHECKOUT_SERVICE = 'CheckoutService';
export const CHECKOUT_PAYMENT_SERVICE = 'CheckoutPaymentService';
export const CHECKOUT_DELIVERY_SERVICE = 'CheckoutDeliveryService';
export const CART_DATA_SERVICE = 'CartDataService';
export const CART_SERVICE = 'CartService';
export const ACTIVE_CART_SERVICE = 'ActiveCartService';
export const PROMOTION_SERVICE = 'PromotionService';
export const ORDER_DETAILS_SERVICE = 'OrderDetailsService';
export const CHECKOUT_LOGIN_COMPONENT = 'CheckoutLoginComponent';
export const AUTH_REDIRECT_SERVICE = 'AuthRedirectService';
export const CHECKOUT_DETAILS_SERVICE = 'CheckoutDetailsService';
export const NOT_CHECKOUT_AUTH_GUARD = 'NotCheckoutAuthGuard';
export const ROUTING_SERVICE = 'RoutingService';
export const SHIPPING_ADDRESS_COMPONENT = 'ShippingAddressComponent';
export const CHECKOUT_CONFIG_SERVICE = 'CheckoutConfigService';
export const TRANSLATION_SERVICE = 'TranslationService';
export const CHECKOUT_PAGE_META_RESOLVER = 'CheckoutPageMetaResolver';
export const ADD_TO_CART_COMPONENT = 'AddToCartComponent';
export const MODAL_SERVICE = 'ModalService';
export const CURRENT_PRODUCT_SERVICE = 'CurrentProductService';
export const CART_NOT_EMPTY_GUARD = 'CartNotEmptyGuard';
export const CART_TOTALS_COMPONENT = 'CartTotalsComponent';
export const MINI_CART_COMPONENT = 'MiniCartComponent';
export const CHECKOUT_ORDER_SUMMARY_COMPONENT = 'CheckoutOrderSummaryComponent';
export const CHECKOUT_PROGRESS_MOBILE_TOP_COMPONENT =
  'CheckoutProgressMobileTopComponent';
export const CHECKOUT_CONFIG = 'CheckoutConfig';
export const ROUTING_CONFIG_SERVICE = 'RoutingConfigService';
export const PAYMENT_METHOD_COMPONENT = 'PaymentMethodComponent';
export const USER_PAYMENT_SERVICE = 'UserPaymentService';
export const GLOBAL_MESSAGE_SERVICE = 'GlobalMessageService';
export const CHECKOUT_AUTH_GUARD = 'CheckoutAuthGuard';
export const CART_PAGE_LAYOUT_HANDLER = 'CartPageLayoutHandler';
export const SELECTIVE_CART_SERVICE = 'SelectiveCartService';
export const CDS_SPARTACUS_EVENT_SERVICE = 'SpartacusEventService';
export const CONSENT_SERVICE = 'ConsentService';
export const CDS_CONFIG = 'CdsConfig';
export const ADDRESS_BOOK_COMPONENT_SERVICE = 'AddressBookComponentService';
export const CHECKOUT_GUARD = 'CheckoutGuard';
export const EXPRESS_CHECKOUT_SERVICE = 'ExpressCheckoutService';
export const CONSENT_MANAGEMENT_FORM_COMPONENT =
  'ConsentManagementFormComponent';
/***** Classes end *****/

/***** Removed public api start *****/
export const CART_EFFECTS = 'CartEffects';
export const WISHLIST_EFFECTS = 'WishlistEffects';
export const CART_VOUCHER_EFFECTS = 'CartVoucherEffects';
export const CART_ENTRY_EFFECTS = 'CartEntryEffects';
export const CART_COMBINED_EFFECTS = 'effects';

export const GET_REDUCERS = 'getReducers';
export const REDUCER_TOKEN = 'reducerToken';
export const REDUCER_PROVIDER = 'reducerProvider';
export const CLEAR_CART_STATE = 'clearCartState';
export const META_REDUCERS = 'metaReducers';
export const CLEAR_MULTI_CART_STATE = 'clearMultiCartState';
export const MULTI_CART_META_REDUCERS = 'multiCartMetaReducers';
export const MULTI_CART_REDUCER_TOKEN = 'multiCartReducerToken';
export const GET_MULTI_CART_REDUCERS = 'getMultiCartReducers';
export const MULTI_CART_REDUCER_PROVIDER = 'multiCartReducerProvider';

export const SITE_CONTEXT_PARAMS_PROVIDERS = 'siteContextParamsProviders';
export const INIT_SITE_CONTEXT_ROUTES_HANDLER = 'initSiteContextRoutesHandler';
export const INITITIALIZE_CONTEXT = 'inititializeContext';
export const CONTEXT_SERVICE_PROVIDERS = 'contextServiceProviders';
/***** Removed public api end *****/

/***** Properties start *****/
export const CMS_COMPONENT_DATA_PROPERTY_NAME = 'componentData';
/***** Properties end *****/

/***** APIs start *****/
export const TODO_SPARTACUS = 'TODO:Spartacus -';

export const GET_COMPONENT_STATE_OLD_API = 'getComponentState';
export const GET_COMPONENTS_STATE_NEW_API = 'getComponentsState';

export const GET_COMPONENT_ENTITIES_OLD_API = 'getComponentEntities';
export const GET_COMPONENT_ENTITIES_COMMENT = `// ${TODO_SPARTACUS} '${GET_COMPONENT_ENTITIES_OLD_API}' has been removed, please use some of the newer API methods.`;

export const COMPONENT_STATE_SELECTOR_FACTORY_OLD_API =
  'componentStateSelectorFactory';
export const COMPONENTS_STATE_SELECTOR_FACTORY_NEW_API =
  'componentsLoaderStateSelectorFactory';

export const COMPONENT_SELECTOR_FACTORY_OLD_API = 'componentSelectorFactory';
export const COMPONENTS_SELECTOR_FACTORY_NEW_API = 'componentsSelectorFactory';
/***** APIs end *****/

/***** Misc start *****/
export const CSS_DOCS_URL =
  'https://sap.github.io/cloud-commerce-spartacus-storefront-docs/deprecation-guide/css';
/***** Misc end *****/
