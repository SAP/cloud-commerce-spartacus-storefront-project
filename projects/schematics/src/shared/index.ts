export * from './constants';
export * from './feature-libs-constants';
export * from './utils/index';
