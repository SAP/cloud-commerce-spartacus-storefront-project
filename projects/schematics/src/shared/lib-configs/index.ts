export * from './asm-schematics-config';
export * from './cart-schematics-config';
export * from './checkout-schematics-config';
export * from './order-schematics-config';
export * from './user-schematics-config';
