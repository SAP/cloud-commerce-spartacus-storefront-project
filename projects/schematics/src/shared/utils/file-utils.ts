import { experimental, strings } from '@angular-devkit/core';
import { SchematicsException, Tree } from '@angular-devkit/schematics';
import { getProjectTargetOptions } from '@angular/cdk/schematics';
import { parseTsconfigFile } from '@angular/core/schematics/utils/typescript/parse_tsconfig';
import { getSourceNodes } from '@schematics/angular/utility/ast-utils';
import {
  Change,
  InsertChange,
  ReplaceChange,
} from '@schematics/angular/utility/change';
import { dirname, relative } from 'path';
import * as ts from 'typescript';

export function getTsSourceFile(tree: Tree, path: string): ts.SourceFile {
  const buffer = tree.read(path);
  if (!buffer) {
    throw new SchematicsException(`Could not read file (${path}).`);
  }
  const content = buffer.toString();
  const source = ts.createSourceFile(
    path,
    content,
    ts.ScriptTarget.Latest,
    true
  );

  return source;
}

export function getAllTsSourceFiles(
  tsconfigPath: string,
  tree: Tree,
  basePath: string
): ts.SourceFile[] {
  const parsed = parseTsconfigFile(tsconfigPath, dirname(tsconfigPath));
  const host = createMigrationCompilerHost(tree, parsed.options, basePath);
  const program = ts.createProgram(parsed.fileNames, parsed.options, host);
  return program
    .getSourceFiles()
    .filter(
      f => !f.isDeclarationFile && !program.isSourceFileFromExternalLibrary(f)
    );
}

export function getIndexHtmlPath(
  project: experimental.workspace.WorkspaceProject
): string {
  const buildOptions = getProjectTargetOptions(project, 'build');

  if (!buildOptions.index) {
    throw new SchematicsException('"index.html" file not found.');
  }

  return buildOptions.index;
}

export function getPathResultsForFile(
  tree: Tree,
  file: string,
  directory?: string
): string[] {
  const results: string[] = [];
  const dir = directory || '/';

  tree.getDir(dir).visit(filePath => {
    if (filePath.endsWith(file)) {
      results.push(filePath);
    }
  });

  return results;
}

export enum InsertDirection {
  LEFT,
  RIGHT,
}

export function commitChanges(
  host: Tree,
  path: string,
  changes: Change[] | null,
  insertDirection: InsertDirection
): void {
  if (!changes) {
    return;
  }

  const recorder = host.beginUpdate(path);
  changes.forEach(change => {
    if (change instanceof InsertChange) {
      const pos = change.pos;
      const toAdd = change.toAdd;
      if (insertDirection === InsertDirection.LEFT) {
        recorder.insertLeft(pos, toAdd);
      } else {
        recorder.insertRight(pos, toAdd);
      }
    } else if (change instanceof ReplaceChange) {
      const pos = change['pos'];
      const oldText = change['oldText'];
      const newText = change['newText'];

      recorder.remove(pos, oldText.length);
      if (insertDirection === InsertDirection.LEFT) {
        recorder.insertLeft(pos, newText);
      } else {
        recorder.insertRight(pos, newText);
      }
    }
  });
  host.commitUpdate(recorder);
}

export function defineProperty(
  nodes: ts.Node[],
  path: string,
  toAdd: string
): InsertChange {
  const constructorNode = nodes.find(n => n.kind === ts.SyntaxKind.Constructor);

  if (!constructorNode) {
    throw new SchematicsException(`No constructor found in ${path}.`);
  }

  return new InsertChange(path, constructorNode.pos + 1, toAdd);
}

export function injectService(
  nodes: ts.Node[],
  path: string,
  serviceName: string,
  propertyName?: string
): InsertChange {
  const constructorNode = nodes.find(n => n.kind === ts.SyntaxKind.Constructor);

  if (!constructorNode) {
    throw new SchematicsException(`No constructor found in ${path}.`);
  }

  const parameterListNode = constructorNode
    .getChildren()
    .find(n => n.kind === ts.SyntaxKind.SyntaxList);

  if (!parameterListNode) {
    throw new SchematicsException(
      `No no parameter list found in ${path}'s constructor.`
    );
  }

  propertyName = propertyName
    ? strings.camelize(propertyName)
    : strings.camelize(serviceName);
  const toAdd = `private ${propertyName}: ${strings.classify(serviceName)}`;
  return new InsertChange(path, parameterListNode.pos, toAdd);
}

export function insertCommentAboveIdentifier(
  sourcePath: string,
  source: ts.SourceFile,
  identifierName: string,
  comment: string
): InsertChange[] {
  const callExpressionNodes = findNodesByTextAndKind(
    source,
    identifierName,
    ts.SyntaxKind.Identifier
  );
  const changes: InsertChange[] = [];
  callExpressionNodes.forEach(n =>
    changes.push(
      new InsertChange(
        sourcePath,
        getLineStartFromTSFile(source, n.getFullStart()),
        comment
      )
    )
  );
  return changes;
}

export function renameIdentifierNode(
  sourcePath: string,
  source: ts.SourceFile,
  oldName: string,
  newName: string
): ReplaceChange[] {
  const callExpressionNodes = findNodesByTextAndKind(
    source,
    oldName,
    ts.SyntaxKind.Identifier
  );
  const changes: ReplaceChange[] = [];
  callExpressionNodes.forEach(n =>
    changes.push(new ReplaceChange(sourcePath, n.getStart(), oldName, newName))
  );
  return changes;
}

function findNodesByTextAndKind(
  source: ts.SourceFile,
  text: string,
  syntaxKind: ts.SyntaxKind
): ts.Node[] {
  const nodes = getSourceNodes(source);
  return nodes
    .filter(n => n.kind === syntaxKind)
    .filter(n => n.getText() === text);
}

function getLineStartFromTSFile(
  source: ts.SourceFile,
  position: number
): number {
  const lac = source.getLineAndCharacterOfPosition(position);
  const lineStart = source.getPositionOfLineAndCharacter(lac.line, 0);

  return lineStart;
}

// as this is copied from https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/app-shell/index.ts#L211, no need to test Angular's code
export function getMetadataProperty(
  metadata: ts.Node,
  propertyName: string
): ts.PropertyAssignment {
  const properties = (metadata as ts.ObjectLiteralExpression).properties;
  const property = properties
    .filter(prop => prop.kind === ts.SyntaxKind.PropertyAssignment)
    .filter((prop: ts.PropertyAssignment) => {
      const name = prop.name;
      switch (name.kind) {
        case ts.SyntaxKind.Identifier:
          return (name as ts.Identifier).getText() === propertyName;
        case ts.SyntaxKind.StringLiteral:
          return (name as ts.StringLiteral).text === propertyName;
      }

      return false;
    })[0];

  return property as ts.PropertyAssignment;
}

// copied from https://github.com/angular/angular/blob/master/packages/core/schematics/utils/typescript/compiler_host.ts#L12, no need to test angular's code
export function createMigrationCompilerHost(
  tree: Tree,
  options: ts.CompilerOptions,
  basePath: string,
  fakeRead?: (fileName: string) => string | null
): ts.CompilerHost {
  const host = ts.createCompilerHost(options, true);

  // We need to overwrite the host "readFile" method, as we want the TypeScript
  // program to be based on the file contents in the virtual file tree. Otherwise
  // if we run multiple migrations we might have intersecting changes and
  // source files.
  host.readFile = fileName => {
    const treeRelativePath = relative(basePath, fileName);
    const fakeOutput = fakeRead ? fakeRead(treeRelativePath) : null;
    const buffer =
      fakeOutput === null ? tree.read(treeRelativePath) : fakeOutput;
    // Strip BOM as otherwise TSC methods (Ex: getWidth) will return an offset,
    // which breaks the CLI UpdateRecorder.
    // See: https://github.com/angular/angular/pull/30719
    return buffer ? buffer.toString().replace(/^\uFEFF/, '') : undefined;
  };

  return host;
}
