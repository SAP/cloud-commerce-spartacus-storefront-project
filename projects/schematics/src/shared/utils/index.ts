export {
  createNewConfig,
  getConfig,
  getExistingStorefrontConfigNode,
  mergeConfig,
} from './config-utils';
export {
  commitChanges,
  findMultiLevelNodesByTextAndKind,
  getTsSourceFile,
  InsertDirection,
  removeImport,
} from './file-utils';
export {
  addImport,
  addToMetadata,
  addToModuleImports,
  addToModuleImportsAndCommitChanges,
  addToModuleProviders,
  createImportChange,
} from './module-file-utils';
export {
  getSpartacusSchematicsVersion,
  readPackageJson,
} from './package-utils';
export {
  getDefaultProjectNameFromWorkspace,
  getProjectTargets,
  getSourceRoot,
  getWorkspace,
} from './workspace-utils';
