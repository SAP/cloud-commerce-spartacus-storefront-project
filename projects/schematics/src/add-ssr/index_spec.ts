import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import { Style } from '@angular/cli/lib/config/schema';
import { Schema as ApplicationOptions } from '@schematics/angular/application/schema';
import * as path from 'path';
import { UTF_8 } from '../shared/constants';
import { getPathResultsForFile } from '../shared/utils/file-utils';

const collectionPath = path.join(__dirname, '../collection.json');

describe('add-ssr', () => {
  const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

  let appTree: UnitTestTree;

  const workspaceOptions: any = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const defaultOptions = {
    project: 'schematics-test',
    target: 'build',
    configuration: 'production',
    baseSite: 'electronics',
    baseUrl: 'https://localhost:9002',
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync('add-spartacus', defaultOptions, appTree)
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync('add-ssr', defaultOptions, appTree)
      .toPromise();
  });

  describe('package.json', () => {
    it('should add SSR dependencies', async () => {
      const packageJson = appTree.readContent('/package.json');
      const packageObj = JSON.parse(packageJson);
      const depPackageList = Object.keys(packageObj.dependencies);

      expect(depPackageList.includes('@angular/platform-server')).toBe(true);
      expect(depPackageList.includes('@nguniversal/express-engine')).toBe(true);
    });

    it('should contain additional build scripts', async () => {
      const buffer = appTree.read('package.json');
      expect(buffer).toBeTruthy();
      if (buffer) {
        const packageJsonFileObject = JSON.parse(buffer.toString(UTF_8));
        expect(packageJsonFileObject.scripts['build:ssr']).toBeTruthy();
        expect(packageJsonFileObject.scripts['serve:ssr']).toBeTruthy();
      }
    });
  });

  describe('angular.json', () => {
    it('should contain changed output path', async () => {
      const buffer = appTree.read('angular.json');
      expect(buffer).toBeTruthy();
      if (buffer) {
        const angularJsonFileObject = JSON.parse(buffer.toString(UTF_8));
        const projectArchitectObject =
          angularJsonFileObject.projects[defaultOptions.project].architect.build
            .options;
        expect(projectArchitectObject['outputPath']).toEqual(
          `dist/${defaultOptions.project}`
        );
      }
    });
    it('should contain server build configuration', async () => {
      const buffer = appTree.read('angular.json');
      expect(buffer).toBeTruthy();
      if (buffer) {
        const angularJsonFileObject = JSON.parse(buffer.toString(UTF_8));
        const projectArchitectObject =
          angularJsonFileObject.projects[defaultOptions.project].architect;
        expect(projectArchitectObject['server']).toBeTruthy();
      }
    });
  });

  describe('app.server.module.ts', () => {
    it('should contain ServerTransferStateModule import', async () => {
      const appServerModulePath = getPathResultsForFile(
        appTree,
        'app.server.module.ts',
        '/src'
      )[0];
      const buffer = appTree.read(appServerModulePath);
      expect(buffer).toBeTruthy();
      if (buffer) {
        const appServerModule = buffer.toString(UTF_8);
        expect(
          appServerModule.includes('ServerTransferStateModule')
        ).toBeTruthy();
        expect(
          appServerModule.includes('@angular/platform-server')
        ).toBeTruthy();
      }
    });
  });

  describe('app.server.module.ts', () => {
    it('should contain ServerTransferStateModule import', async () => {
      const indexHtmlPath = getPathResultsForFile(
        appTree,
        'index.html',
        '/src'
      )[0];
      const buffer = appTree.read(indexHtmlPath);
      expect(buffer).toBeTruthy();
      if (buffer) {
        const indexHtmlFile = buffer.toString(UTF_8);
        expect(
          indexHtmlFile.includes('meta name="occ-backend-base-url"')
        ).toBeTruthy();
      }
    });
  });

  describe('webpack.server.config.js', () => {
    it('should be created', async () => {
      const buffer = appTree.read('./webpack.server.config.js');
      expect(buffer).toBeTruthy();
      if (buffer) {
        const webpackConfigFile = buffer.toString(UTF_8);
        expect(webpackConfigFile.length).toBeGreaterThan(0);
      }
    });
  });

  describe('main.server.ts', () => {
    it('should be modified with custom ngExpressImport and export', async () => {
      const buffer = appTree.read('./src/main.server.ts');
      expect(buffer).toBeTruthy();
      if (buffer) {
        const mainServerFileString = buffer.toString(UTF_8);
        expect(mainServerFileString.length).toBeGreaterThan(0);

        expect(
          mainServerFileString.includes(
            'export { ngExpressEngine } from "@nguniversal/express-engine";'
          )
        ).toBeFalsy();

        expect(
          mainServerFileString.includes(
            "import { ngExpressEngine as engine } from '@nguniversal/express-engine';"
          )
        ).toBeTruthy();

        expect(
          mainServerFileString.includes(
            "import { NgExpressEngineDecorator } from '@spartacus/core';"
          )
        ).toBeTruthy();

        expect(
          mainServerFileString.includes(
            'export const ngExpressEngine = NgExpressEngineDecorator.get(engine);'
          )
        ).toBeTruthy();
      }
    });
  });
});
