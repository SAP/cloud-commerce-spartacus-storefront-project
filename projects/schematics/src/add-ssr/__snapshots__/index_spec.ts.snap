// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add-ssr app.module.server.ts should be updated 1`] = `
"import { NgModule } from '@angular/core';
import { ServerModule } from '@angular/platform-server';

import { AppModule } from './app.module';
import { AppComponent } from './app.component';
import { provideServer } from '@spartacus/setup/ssr';

@NgModule({
  imports: [
    AppModule,
    ServerModule,
  ],
  bootstrap: [AppComponent],
  providers: [
    ...provideServer({
       serverRequestOrigin: process.env['SERVER_REQUEST_ORIGIN'],
     }),
  ],
})
export class AppServerModule {}
"
`;

exports[`add-ssr index.html should contain occ-backend-base-url attribute in meta tags 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>SchematicsTest</title>
  <base href="/">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="icon" type="image/x-icon" href="favicon.ico">
  <meta name="occ-backend-base-url" content="https://localhost:9002" />
</head>
<body>
  <app-root></app-root>
</body>
</html>
"
`;

exports[`add-ssr server.ts should be configured properly 1`] = `
"import { APP_BASE_HREF } from '@angular/common';
import {
  NgExpressEngineDecorator,
  ngExpressEngine as engine,
} from '@spartacus/setup/ssr';

import express from 'express';
import { join } from 'path';

import { existsSync } from 'node:fs';
import AppServerModule from './src/main.server';

const ngExpressEngine = NgExpressEngineDecorator.get(engine);

// The Express app is exported so that it can be used by serverless Functions.
export function app() {
  const server = express();
  const distFolder = join(process.cwd(), 'dist/schematics-test/browser');
  const indexHtml = existsSync(join(distFolder, 'index.original.html'))
    ? 'index.original.html'
    : 'index';

  server.set('trust proxy', 'loopback');

  server.engine(
    'html',
    ngExpressEngine({
      bootstrap: AppServerModule,
    })
  );

  server.set('view engine', 'html');
  server.set('views', distFolder);

  // Serve static files from /browser
  server.get(
    '*.*',
    express.static(distFolder, {
      maxAge: '1y',
    })
  );

  // All regular routes use the Universal engine
  server.get('*', (req, res) => {
    res.render(indexHtml, {
      req,
      providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
  });

  return server;
}

function run() {
  const port = process.env['PORT'] || 4000;

  // Start up the Node server
  const server = app();
  server.listen(port, () => {
    console.log(\`Node Express server listening on http://localhost:\${port}\`);
  });
}

run();
"
`;
